logging: net: runtime setting of syslog server ip address

dcorbeil opened the pull request:
Added functionality to set the syslog server's ip address at runtime 
as well as sample for syslog network backend.  This will be useful for 
setting the server's ip address when it is not known at build time or in
a dynamic network.

rlubos commented on subsys/logging/backends/log_backend_net.c:
```diff
@@ -196,20 +197,31 @@ static int format_set(const struct log_backend *const backend, uint32_t log_type
+{
+	net_sin(&server_addr)->sin_port = htons(514);
+
+	bool ret = net_ipaddr_parse(addr, strlen(addr), &server_addr);
```
The thing with runtime address configuration is that with the current implementation it will have no effect if the corresponding net_context has already been initialized (see `do_net_init()` and `net_context_connect()` in there).

I think that at the very least, the function should return an error in such case (`net_init_done` is set). Perhaps a better solution though would be to release the current `net_context` and clear `net_init_done` flag so that the log backend can create a new one, with a new destination address. WDYT?

dcorbeil replied:
It'll for sure make this feature much more flexible and adding flexibility was my initial motivation. I'll add a way to change the server address on the fly.

rlubos reviewed (changes requested):
Ah, sorry I didn't mean to approve yet...

dcorbeil commented on include/zephyr/logging/log_backend_net.h:
```diff
@@ -0,0 +1,29 @@
+ * @details This function allows the user to set an IP address at runtime before the syslog network
+ *          backend is initialized since the string address gets parsed in log_backend_api::init().
+ *          For this to work, CONFIG_LOG_BACKEND_NET_SERVER_RUNTIME needs to be enabled and is
+ *          mutually exclusive with CONFIG_LOG_BACKEND_NET_SERVER_STATIC.
```
Yes will do. That makes sense.

rlubos commented on samples/net/syslog_net/runtime-srv-addr.conf:
```diff
@@ -0,0 +1,2 @@
+CONFIG_LOG_BACKEND_NET_SERVER_RUNTIME=y
+CONFIG_LOG_BACKEND_NET_SERVER=""
```
Empty line needed at the end of the file.

rlubos commented:
Plus, I think the commits could be squashed.

dcorbeil commented on subsys/logging/backends/Kconfig.net:
```diff
@@ -15,6 +15,15 @@ config LOG_BACKEND_NET
 
 if LOG_BACKEND_NET
 
+config LOG_BACKEND_NET_SERVER_RUNTIME
```
I was using it mostly for the sample app so that it was easier and more obvious to test and show future users what where the various capabilities of the `syslog_net` backend. But I can remove it no problem.

dcorbeil commented on include/zephyr/logging/log_backend_net.h:
```diff
@@ -0,0 +1,31 @@
+#ifdef CONFIG_LOG_BACKEND_NET_SERVER_RUNTIME
+bool log_backend_net_set_addr(struct log_backend const *const backend, const char *addr);
+#endif
+
```
Yes

dcorbeil commented on include/zephyr/logging/log_backend_net.h:
```diff
@@ -0,0 +1,31 @@
+ */
+
+#ifdef CONFIG_LOG_BACKEND_NET_SERVER_RUNTIME
+bool log_backend_net_set_addr(struct log_backend const *const backend, const char *addr);
```
Good point will remove.

rlubos commented on include/zephyr/logging/log_backend_net.h:
```diff
@@ -0,0 +1,28 @@
+ *          the new address next time process() gets called.
+ *
+ * @param addr     String that contains the IP address.
+ * @param backend  Pointer to the backend instance.
```
The docs for `backend` parameter should also be removed.

rlubos commented on include/zephyr/logging/log_backend_net.h:
```diff
@@ -0,0 +1,28 @@
+ *
+ * @return True if parsing could be done, false otherwise.
+ */
+
```
Unneeded empty line.

dcorbeil replied:
Yes

rlubos commented on include/zephyr/logging/log_backend_net.h:
```diff
@@ -0,0 +1,28 @@
+#define ZEPHYR_LOG_BACKEND_NET_H_
+
+#include <stdbool.h>
+
```
Would make sense to add cpp guard here:
```c
#ifdef __cplusplus
extern "C" {
#endif

...

#ifdef __cplusplus
}
#endif
```

dcorbeil replied:
Yes

rlubos commented on samples/net/syslog_net/runtime-srv-addr.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_LOG_BACKEND_NET_SERVER=""
```
The overlay file should also disable ` CONFIG_LOG_BACKEND_NET_AUTOSTART=n` if I'm not mistaken, the runtime address setting will only take place if autostart is disabled.

Also regarding the filename, please use `overlay-runtime-srv-addr.conf` so it's clear we deal with an overlay file.

dcorbeil replied:
Will do

rlubos reviewed (commented):
Thanks, just some final nits from my side, looks good otherwise.

jukkar commented on samples/net/syslog_net/src/main.c:
```diff
@@ -36,6 +37,16 @@ int main(void)
+			   address.
+			 */
+			if (strlen(CONFIG_LOG_BACKEND_NET_SERVER) == 0) {
+				log_backend_net_set_addr("192.0.2.2:514");
```
Please no hardcoded addresses as it makes it impossible to change the address without editing the file. Please add an option in sample Kconfig file if we need a default value to be used.

jukkar commented on samples/net/syslog_net/Kconfig:
```diff
@@ -12,4 +12,20 @@ config NET_SAMPLE_SEND_ITERATIONS
 	  zero means that the defaults in the application are used.
 
+
+config NET_SAMPLE_NET_SERVER_RUNTIME
```
The latter NET seems uncessary, `NET_SAMPLE_SERVER_RUNTIME` should be enough here.

jukkar reviewed (commented):
LGTM, minor nit about the option naming.

Commit:
logging: runtime setting of syslog server ip address

Added functionality to change the syslog server's
ip address at runtime as well as sample for
syslog network backend

Signed-off-by: David Corbeil <david.corbeil@dynon.com>
